replica/backup.go

```diff
@@ -175,24 +180,92 @@ func (rb *Backup) findIndex(id string) int {
 	return -1
 }
 
+type Hole struct {
+	f      types.DiffDisk
+	offset int64
+	len    int64
+}
+
+var HoleCreatorChan = make(chan Hole, 1000000)
+
+//CreateHoles removes the offsets from corresponding sparse files
+func CreateHoles() error {
+	retryCount := 0
+	for {
+		hole := <-HoleCreatorChan
+	retry:
+		if err := syscall.Fallocate(int(hole.f.Fd()),
+			sparse.FALLOC_FL_KEEP_SIZE|sparse.FALLOC_FL_PUNCH_HOLE,
+			hole.offset*defaultSectorSize, hole.len*4096); err != nil {
+			logrus.Errorf("ERROR in creating hole: %v, Retry_Count: %v", err, retryCount)
+			time.Sleep(1)
+			retryCount++
+			if retryCount == 5 {
+				logrus.Fatalf("Error Creating holes: %v", err)
+			}
+			goto retry
+		}
+		retryCount = 0
+	}
+}
+
+func sendToCreateHole(f types.DiffDisk, offset int64, len int64) {
+	hole := Hole{
+		f:      f,
+		offset: offset,
+		len:    len,
+	}
+	HoleCreatorChan <- hole
+}
+
+//Preload creates a mapping of block number to fileIndx (d.location).
+//This is done with the help of extent list fetched from filesystem.
+//Extents list in each file is traversed and the location table is updated
 func preload(d *diffDisk) error {
+	var file types.DiffDisk
+	var length int64
+	var lOffset int64
+	var fileIndx byte
+	var userCreatedSnapIndx byte
 	for i, f := range d.files {
 		if i == 0 {
 			continue
 		}
 		if i == 1 {
+			//TODO Check if this zeroing is needed
 			// Reinitialize to zero so that we can detect holes in the base snapshot
 			for j := 0; j < len(d.location); j++ {
 				d.location[j] = 0
 			}
 		}
+		if d.UserCreatedSnap[i] {
+			userCreatedSnapIndx = byte(i)
+		}
 		generator := newGenerator(d, f)
 		for offset := range generator.Generate() {
+			if d.location[offset] != 0 {
+				if d.files[d.location[offset]] != file {
+					if file != nil && fileIndx > userCreatedSnapIndx && !d.ReadOnlyIndx[fileIndx] {
```

Comment:
file and fileIndx are not initialized, and in the loop, it will take previous loop contents.
So, after punching holes to 'file1', and, if any number of initial extents in next file i.e., 'file2' matches with 'file1', they won't get unmapped.
And, a comment would be helpful to understand code

Reply:
got to understand that we are trying to find continuous blocks

Reply:
but, still, being file not initialized or initialized to NULL by default, it gets into 'else' case where 'length' is just increased.
Seems to be something wrong here.

Reply:
same file has been verified.. but, contiguity of blocks also need to be verified

Reply:
is it better to check of ReadOnlyIndx while doing the syscall for unmap?